Re-add Synchronize (rsync) module (with improvements)

benosman opened the pull request:
There has been a discussion of the ansible group about problems with the synchronize module which was removed.

https://groups.google.com/d/msg/ansible-project/DTIjx8aAo_4/wTVHAiOAwhYJ

The first commit restores the code from the initial pull request: https://github.com/ansible/ansible/pull/3173

The second commit includes my changes which:
- Attempt to fix the issues in the linked topic: Namely it doesn't work well on the localhost. The reason for this is that in the setup function it was force adding the delegate_to option but was not setting the transport method. This meant that ansible tried to ssh into localhost, which may work on some systems, but for many it doesn't. I used inject['ansible_connection'] = 'local'  to remedy this situation.
- Improved the src and dest handling, and introduced a mode variable for the class. This attempts to determine the mode that the action is being run in. It even takes care of using sudo.
  - The default is remote mode, and operates like copy action, where the src is a local filepath and dest is a remote filepath (if no hosts are specified).
  - If run using local_action or delegate_to is set to 127.0.0.1, then this is local mode. If no hosts are set in src and dest parameters then local filepaths are assumed.
  - If delegate_to is another host, then other mode is used, which uses filepaths on the specified host.
- Added a whole load of extra parameters based on rsync flags, and changed the default behaviour of using the -archive flag, which i found inconvenient as it keeps the owners of files the same as on the local system which if that user doesn't exist on the host then its not ideal. 
- Added change detection - now the response is parsed to determine whether there were any changes or not. Also added dry_run parameter which can be combined with change detection to test first whether there are changes.

I've been testing this for a few days, but would really welcome some further testing.

benosman committed:
Revert "Removing synchronize momentarily as it looks like this won't make 1.3 dates"

This reverts commit dacd311474c0480620625b9f7cc56166333086fa.

benosman committed:
Improve functionality of rsync module.

tima commented:
Oh... inject['ansible_connection'] = 'local' ! [slaps head]

Will try to look this over this evening.

tima commented:
First off, thanks for figuring out what was going wrong for some users. I didn't think of the connection param specifically in inject dictionary. I'm still concerned that artifacts of the original host (before the delegate override) is still hanging out in there and could mess things up under certain conditions. Could being the operative word -- I've yet to see it. I'm just putting it out there. Nevertheless...

I was having a bit of trouble understanding the logic behind some of your changes because the documentation examples weren't updated. 

Did you implement a src_host and dest_host and if so why? I worked pretty hard to apply convention of configuration to the design of this module. These params (and a few others I'll cover later) break that and it makes me sad for whatever that is worth. I'm also not fond of letting users override the src_host and/or dest_host with any arbitrary host rather than the inventory_hostname in scope or a specific delegate. Seems very un-Ansible-like to me. I just can't think of a reason you'd need to specify a host that isn't localhost, the inventory_hostname or a delegate.With those gone the logic of the module becomes more streamlined. 

I understand that you found the defaults of the --archive switch being inconvenient to you (I actually share in this problem since we cannot alter timestamps on a lot of our system) so  I do see the case for switching off some of the options archive encapsulates. I really don't like all the defaults you chose for all the params. It seems like it going against the wisdom of the rsync developers -- I rarely see anyone run rsync without -a that it may as well be a default. I suppose there is a case for not using --archive though that it should be optional rather than hard coded as I had it. However, we should not arbitrarily create our own version of --archive by (dis)enabling the various individual options that one encapsulates. Besides messing with the rsync gods, it makes the sync action declarations more verbose. For instance in my company where rsync can (and want to) set permissions and groups, but not timestamps I would have to do this with how you've implemented things:

```
 command synchronize src=src/ dest=dest/ times=no groups=yes perms=yes
```

Where it could just be:

```
 command synchronize src=src/ dest=dest/ times=no
```

Where --archive will be used and times=no will append --no-times to the rsync call. (I checked and any --no-[OPTION] overrides anything the --archive switch enables.) I think setting by subtraction rather than addition will result in a lot less setting of options in each task than the other way around. So if we really want to run the sync without --archive we could do:

```
 command synchronize src=src/ dest=dest/ archive=no
```

And if you still really wanted to set each option individually you still could:

```
 command synchronize src=src/ dest=dest/ archive=no times=yes recursive=yes links=yes
```

The system you've implemented forces a user to configure how to synchronize rather than apply convention and let users override the behavior if they have a reason. 

All this aside if I did something like this with your implementation what would happen?

```
 command synchronize src=src/ dest=dest/ archive=yes times=no
```

I'm pretty sure I wouldn't get the result I asked for. Seems like a whole in the implementation of those options.

The --dry-run option should be controlled by Ansible's --check mode and not an option of an individual task IMO. Why would I explicitly and immutably do a rsync --dry-run in a playbook? I couldn't think of a reason. 

There's a handful of other specific comments I had that I'm going to attach to the line of code that it pertains to.

benosman commented:
Thanks for such a detailed response.

Firstly, I have to say that most of the work on this pull request was created for my own purposes so I had a working module for my project. Then after fixing the local connection issue i thought it good to contribute back, so some of the changes may be skewed to my requirements rather than what i think the defaults should be.

> Did you implement a src_host and dest_host and if so why? I worked pretty hard to apply convention of configuration to the design of this module. These params (and a few others I'll cover later) break that and it makes me sad for whatever that is worth. I'm also not fond of letting users override the src_host and/or dest_host with any arbitrary host rather than the inventory_hostname in scope or a specific delegate. Seems very un-Ansible-like to me. I just can't think of a reason you'd need to specify a host that isn't localhost, the inventory_hostname or a delegate.With those gone the logic of the module becomes more streamlined.

I created them initially when I was trying to override the host logic that was there before and wasn't working for me. It was a quick fix really and once I refined the host logic they became pretty redundant. I will remove those.

> I understand that you found the defaults of the --archive switch being inconvenient to you (I actually share in this problem since we cannot alter timestamps on a lot of our system) so I do see the case for switching off some of the options archive encapsulates. I really don't like all the defaults you chose for all the params. It seems like it going against the wisdom of the rsync developers -- I rarely see anyone run rsync without -a that it may as well be a default. I suppose there is a case for not using --archive though that it should be optional rather than hard coded as I had it. However, we should not arbitrarily create our own version of --archive by (dis)enabling the various individual options that one encapsulates. Besides messing with the rsync gods, it makes the sync action declarations more verbose. For instance in my company where rsync can (and want to) set permissions and groups, but not timestamps I would have to do this with how you've implemented things:

You raise some good points, and as mentioned above it was created with my own requirements in mind, then put forward as it may be useful. I am certainly open to discussion on this, and when I get a moment I will also give it more thought. Would also appreciate others input on this.

> The --dry-run option should be controlled by Ansible's --check mode and not an option of an individual task IMO. Why would I explicitly and immutably do a rsync --dry-run in a playbook? I couldn't think of a reason.

The reason for this was that i wanted a non invasive way to check whether there was any changes to a remote directory. I build a lot of my config locally and then syncronize it - i hadn't used the --check mode before so this makes sense to me.

I will reply to some of your line comments individually.

Ben

benosman committed:
Add rsync fixes discussed in pull request

benosman commented:
The commit above addresses most of the points Tim raised, with the exception of the defaults for the parameters, as i'd like to get some consensus on what is the most ansible way of doing it.

Here is a summary of what the --archive flag includes.

```
-r, --recursive recurse into directories

-l, --links copy symlinks as symlinks

-p, --perms preserve permissions

-t, --times preserve modification times

-g, --group preserve group

-o, --owner preserve owner (super-user only)

-D same as --devices --specials

--devices preserve device files (super-user only)

--specials preserve special files
```

Most of these would be what I imagine ansible users would expect, but i'm not sure about the group and owner, as i personally would expect the owner to be by default the ansible user.

For quick reference here is the flags we currently have in this module

``` python
recursive = dict(default='yes', type='bool'),
dirs  = dict(default='no', type='bool'),
links = dict(default='yes', type='bool'),
perms = dict(default='yes', type='bool'),
times = dict(default='yes', type='bool'),
owner = dict(default='no', type='bool'),
group = dict(default='no', type='bool'),
archive = dict(default='no', type='bool'),
delete = dict(default='no', type='bool'),
```

benosman commented:
I wonder if it would be easier if you create your own pull request and have things to your liking?

Feel free to take as much or as little of my changes as required, i just wanted to get the ball rolling on this as you said you didn't have the time for it. 

I think i'd prefer to keep my own copy and make the module how i need it for now.

tima commented:
Fair enough I suppose.
